tensorflow_serving/tools/docker/Dockerfile.devel-mkl

```diff
@@ -0,0 +1,129 @@
+        automake \
+        build-essential \
+        ca-certificates \
+        curl \
+        git \
+        libcurl3-dev \
+        libfreetype6-dev \
+        libpng12-dev \
+        libtool \
+        libzmq3-dev \
+        mlocate \
+        openjdk-8-jdk\
+        openjdk-8-jre-headless \
+        pkg-config \
+        python-dev \
+        software-properties-common \
+        swig \
+        unzip \
+        wget \
+        zip \
+        zlib1g-dev \
+        && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+RUN curl -fSsL -O https://bootstrap.pypa.io/get-pip.py && \
+    python get-pip.py && \
+    rm get-pip.py
+
+RUN pip --no-cache-dir install \
+    grpcio \
+    h5py \
+    keras_applications \
+    keras_preprocessing \
+    mock \
+    numpy \
+    requests
+
+# Set up Bazel
+# Need >= 0.15.0 so bazel compiles work with docker bind mounts.
+ENV BAZEL_VERSION 0.15.0
+WORKDIR /
+RUN mkdir /bazel && \
+    cd /bazel && \
+    curl -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" -fSsL -O https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
+    curl -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" -fSsL -o /bazel/LICENSE.txt https://raw.githubusercontent.com/bazelbuild/bazel/master/LICENSE && \
+    chmod +x bazel-*.sh && \
+    ./bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
+    cd / && \
+    rm -f /bazel/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh
+
+# Download TF Serving sources (optionally at specific commit).
+WORKDIR /tensorflow-serving
+RUN git clone --branch=${TF_SERVING_VERSION_GIT_BRANCH} https://github.com/tensorflow/serving . && \
+    git remote add upstream https://github.com/tensorflow/serving.git && \
+    if [ "${TF_SERVING_VERSION_GIT_COMMIT}" != "head" ]; then git checkout ${TF_SERVING_VERSION_GIT_COMMIT} ; fi
+
+
+FROM base_build as binary_build
+
+# Build, and install TensorFlow Serving
+# Use --copt=-march values to get optimized builds appropriate for the hardware
+#   platform of your choice.
+# For ivy-bridge or sandy-bridge
+# --copt=-march="avx" \
+# For haswell, broadwell, or skylake
+# --copt=-march="avx2" \
+ARG TF_SERVING_BUILD_OPTIONS="--config=mkl --copt=-march=native -c opt"
+RUN echo "Building with build options: ${TF_SERVING_BUILD_OPTIONS}"
+
+ARG TF_SERVING_BAZEL_OPTIONS=""
+RUN echo "Building with Bazel options: ${TF_SERVING_BAZEL_OPTIONS}"
+
+RUN bazel build --color=yes --curses=yes \
+    ${TF_SERVING_BAZEL_OPTIONS} \
+    --verbose_failures \
+    --output_filter=DONT_MATCH_ANYTHING \
+    ${TF_SERVING_BUILD_OPTIONS} \
+    tensorflow_serving/model_servers:tensorflow_model_server && \
+    cp bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server \
+    /usr/local/bin/
+
+# Build and install TensorFlow Serving API
```

Comment:
@gautamvasudevan I added all feedback changes as suggested. I'm checking for `mkldbb_verbose` in docker logs after querying model with restapi call. I found an issue, restapi calls are not using `MKL`, whereas when i tested with `grpc` calls, it's using MKL. 

Figured out that, line`106` is not installing `intel optimized tensorflow` while building wheel, it's using default tensorflow .

Can you please explain me more about above `Tensorflow Serving API`. Is this just a wrapper which uses above built `tensorflow-model-server`?  and does Rest calls uses this API?

Reply:
The TensorFlow Serving API is basically the gRPC protos exported as python. It's a client library (so if you're calling via grpc, you're using this). It has nothing to do with the serving side, however.

tensorflow-model-server is built by bazel hermetically, and doesn't use or require any external dependencies on the machine. So whatever options you're building tensorflow-model-server with should get passed down when it builds tensorflow and statically links it.

I'm not sure why REST vs gRPC calls would make a difference in terms of using MKL, as that just handles deserialization of data and passing it to the model.

Reply:
Thanks Gautham, let me check what i can do.

Reply:
yeah quick question. Does REST API calls register verbose or logs in `docker logs`?

Reply:
No, the choice of endpoint doesn't impact logging, and there's no special extra logging of REST API calls.

You can take a look at the [implementation](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/model_servers/http_rest_api_handler.cc), it's pretty straightforward.

Reply:
After analysis figured out that, it's problem with `saved_model_half_plus_two_cpu` as, this is very small . it doesn't use any of MKL operations by default. @tonyreina Will help me to bind ops to mkl related ops.. @gautamvasudevan Can you please point me to the tensorflow code of `half_plus_two` model.

Reply:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/saved_model/saved_model_half_plus_two.py ?? This is the one we have to use?

Reply:
It looks like half_plus_two just uses a simple equation. Are there any example models that make use of convolutional layers? That would be ideal for testing the MKL stuff.

Reply:
or @gautamvasudevan can we add our own simple example which uses MKL stuff,?all i need to do is a small check while querying, i will add new `def` and check while calling. https://github.com/tensorflow/serving/blob/master/tensorflow_serving/tools/docker/tests/docker_test_lib.sh#L165  . Please let me know.

Reply:
Yes, you can add your own model for testing, just be sure to also include the code to generate it.

Reply:
@gautamvasudevan: @tonyreina made changes to `half_plus_two` model to use `mkl` optimizations. I added that saved model in `servables/tensorflow/testdata` and did related changes. In above comment, you mentioned to `add code to generate it`. Can you please tell me where i should add those code generation changes.

Reply:
Now unit tests are fixed. All changes done
```
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nChw16c out:f32_nhwc,num:1,1x1x100x100,0.0209961
Successfully queries model. Response:\n{
    "predictions": [2.5, 3.0, 4.5
    ]
}
Checking for mkldnn_verbose in logs of container: 4e18ca60a823
Stopping TF ModelServer
4e18ca60a823
Clean up TMPDIR:/tmp/tmp.0xYPqY0qAn
Target //tensorflow_serving/tools/docker/tests:unittest_dockerfile_mkl up-to-date:
  bazel-bin/tensorflow_serving/tools/docker/tests/unittest_dockerfile_mkl
INFO: Elapsed time: 10.112s, Critical Path: 6.94s
INFO: 1 process: 1 local.
INFO: Build completed successfully, 5 total actions
//tensorflow_serving/tools/docker/tests:unittest_dockerfile_mkl          PASSED in 6.9s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see whINFO: Build completed successfully, 5 total actions
```
```
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nChw16c out:f32_nhwc,num:1,1x1x100x100,0.209961
Successfully queries model. Response:\n{
    "predictions": [2.5, 3.0, 4.5
    ]
}
Checking for mkldnn_verbose in logs of container: 9ba89c20ba42
Stopping TF ModelServer
9ba89c20ba42
Clean up TMPDIR:/tmp/tmp.wDC7ZFrJeK
Target //tensorflow_serving/tools/docker/tests:unittest_dockerfile_devel_mkl up-to-date:
  bazel-bin/tensorflow_serving/tools/docker/tests/unittest_dockerfile_devel_mkl
INFO: Elapsed time: 7.398s, Critical Path: 7.04s
INFO: 1 process: 1 local.
INFO: Build completed successfully, 5 total actions
//tensorflow_serving/tools/docker/tests:unittest_dockerfile_devel_mkl    PASSED in 7.0s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see whINFO: Build completed successfully, 5 total actions
```

Reply:
Hmm, I don't see the files added to testdata (or the corresponding changes to the unit test) in this PR. You can add the script that generates the model inside testdata as well.

Reply:
Yes , i haven't pushed the changes yet. Will add script to generate code and then will push.

Reply:
sounds good! If the tests complete in 10s (as it looks like), then the test size can be returned to small so you don't get a warning

Reply:
yeah i changed it `small`.